Correct serialization of Selectors

annevk opened the pull request:
#### 892935d504cf4b301a232d0b4399b5ac97f67900
<pre>
Correct serialization of Selectors
<a href="https://bugs.webkit.org/show_bug.cgi?id=184604">https://bugs.webkit.org/show_bug.cgi?id=184604</a>
rdar://97092572

Reviewed by Antti Koivisto and Tim Nguyen.

We did not use serializeIdentifier for qualified name prefixes and
local names. E.g., x:x was not serialized as x\:x. This was previously
attempted, but it would do the wrong thing for * and thus backed out.
So now we do not escape when prefix or local name is *.

The only edge case remaining is whether the * and \* inputs should mean
the same thing. <a href="https://github.com/w3c/csswg-drafts/issues/8911">https://github.com/w3c/csswg-drafts/issues/8911</a> will
eventually address this.

Meanwhile this seems like a solid improvement.

Also synchronize web-platform-tests/css/cssom up to upstream commit
57c5006103146974adb3af50d325c3b6ce7153d0 for real this time.
264934@main missed a couple changes.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssimportrule-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssimportrule.html:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/selectorSerialize-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/selectorSerialize.html:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/xml-stylesheet-pi-in-doctype.xhtml:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::serializeIdentifierOrStar):
(WebCore::CSSSelector::selectorText const):

Canonical link: <a href="https://commits.webkit.org/264980@main">https://commits.webkit.org/264980@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bc5402768d1f6c7817c479d37a5374208de2ba7f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9340 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9616 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9846 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11000 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9214 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9349 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11602 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9578 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12098 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9489 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10408 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8043 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11157 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7704 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8518 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15960 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8802 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8669 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12005 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9163 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7462 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8376 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/8424 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2262 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12600 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8925 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/selectors/w3c-import.log:
```diff
@@ -219,11 +219,8 @@ List of files:
 /LayoutTests/imported/w3c/web-platform-tests/css/selectors/is-default-ns-001.html
-/LayoutTests/imported/w3c/web-platform-tests/css/selectors/is-default-ns-002-expected.html
 /LayoutTests/imported/w3c/web-platform-tests/css/selectors/is-default-ns-002.html
-/LayoutTests/imported/w3c/web-platform-tests/css/selectors/is-default-ns-003-expected.html
```
Not sure how you ended up deleting these files, but it seems wrong to me, the test wouldn't be able to run. Fwiw, I resynced this directory very recently: https://github.com/WebKit/WebKit/pull/14420

annevk replied:
Yeah it's wrong. Filed https://bugs.webkit.org/show_bug.cgi?id=257674.

mdubet commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -403,15 +403,24 @@ static void outputNthChildAnPlusB(const CSSSelector& selector, StringBuilder& bu
     }
 }
 
+inline void serializeIdentifierOrStar(const AtomString& identifier, StringBuilder& builder)
```
Not sure it's worth a change, but a lambda with `builder` ref-captured would be more self-contained/cleaner (inside `CSSSelector::selectorText`)

annevk replied:
Happy to make this change if other reviewers agree. (Although I'll need to learn a bit about C++ lambdas.)

nt1m replied:
I agree.

nt1m reviewed (approved):
Looks good, please address the lambda comment.

```
auto serializeIdentifierOrStar = [&] (const AtomString& identifier) {
  [...]
};
```

annevk committed:
Correct serialization of Selectors
https://bugs.webkit.org/show_bug.cgi?id=184604
rdar://97092572

Reviewed by Antti Koivisto and Tim Nguyen.

We did not use serializeIdentifier for qualified name prefixes and
local names. E.g., x:x was not serialized as x\:x. This was previously
attempted, but it would do the wrong thing for * and thus backed out.
So now we do not escape when prefix or local name is *.

The only edge case remaining is whether the * and \* inputs should mean
the same thing. https://github.com/w3c/csswg-drafts/issues/8911 will
eventually address this.

Meanwhile this seems like a solid improvement.

Also synchronize web-platform-tests/css/cssom up to upstream commit
57c5006103146974adb3af50d325c3b6ce7153d0 for real this time.
264934@main missed a couple changes.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssimportrule-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssimportrule.html:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/selectorSerialize-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/selectorSerialize.html:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/xml-stylesheet-pi-in-doctype.xhtml:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::serializeIdentifierOrStar):
(WebCore::CSSSelector::selectorText const):

Canonical link: https://commits.webkit.org/264980@main
